### PR TITLE
Switch to Xunit

### PR DESCRIPTION
**Pull request by pdelvo**
Do not merge.

I have some issues with the built in test explorer in VS. It does not run all the tests. 72 of them just stay as "Not Run Tests". The console runner however works fine. Maybe this is an issue with my machine.

I have not updated the appveyor file too.

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1600UnitTests.cs [hunk 1]
```diff
@@ -5,19 +5,19 @@
...
-    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Xunit;
     using StyleCop.Analyzers.DocumentationRules;
     using TestHelper;
 
     /// <summary>
     /// This class contains unit tests for <see cref="SA1600ElementsMustBeDocumented"/>-
     /// </summary>
-    [TestClass]
+    [Collection("FooCollection")]
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1600UnitTests.cs]: :exclamation: Lets remove the `[Collection]` attribute for now.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/MaintainabilityRules/SA1400UnitTests.cs [hunk 2]
```diff
@@ -7,252 +7,251 @@
     using Microsoft.CodeAnalysis;
     using Microsoft.CodeAnalysis.CodeFixes;
     using Microsoft.CodeAnalysis.Diagnostics;
-    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Xunit;
     using TestHelper;
-
-    [TestClass]
+    [Collection("Foo")]
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/MaintainabilityRules/SA1400UnitTests.cs]: :exclamation: Omit this for now.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/MaintainabilityRules/SA1404UnitTests.cs [hunk 3]
```diff
@@ -236,7 +235,7 @@ public string Bar()
             }
             catch (NullReferenceException)
             {
-                Assert.Fail("Diagnostic threw NullReferenceException");
+                Assert.True(false, "Diagnostic threw NullReferenceException");
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/MaintainabilityRules/SA1404UnitTests.cs]: :question: Any idea why this `try`/`catch` block is here at all? Fixing it is outside the scope of this pull request and should be handled separately (or at minimum, in a commit by itself).

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -33,3 +34,6 @@
 [assembly: AssemblyVersion("1.0.0.0")]
 [assembly: AssemblyFileVersion("1.0.0.0")]
 [assembly: AssemblyInformationalVersion("1.0.0.0-dev")]
+
+// XUnit
+[assembly: CollectionBehavior(CollectionBehavior.CollectionPerClass, DisableTestParallelization = true, MaxParallelThreads = 8)]
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/Properties/AssemblyInfo.cs]: :exclamation: This is the cause of the tests not completing. Revert the changes to this file for now and we can consider adding it back later.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1121UnitTests.cs [hunk 5]
```diff
@@ -67,7 +66,7 @@ public class SA1121UnitTests : CodeFixVerifier
                 }
                 catch (Exception ex)
                 {
-                    throw new AssertFailedException("Type failed: " + item.Item1 + Environment.NewLine + ex.Message, ex);
+                    Assert.True(false, "Type failed: " + item.Item1 + Environment.NewLine + ex.Message);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1121UnitTests.cs]: :question: Any idea why this `try`/`catch` block is here? (Same as my previous question)

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/StyleCop.Analyzers.Test.csproj [hunk 6]
```diff
@@ -14,6 +17,7 @@
     <TargetFrameworkVersion>v4.5</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
     <TargetFrameworkProfile />
+    <NuGetPackageImportStamp>b08d3586</NuGetPackageImportStamp>
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/StyleCop.Analyzers.Test.csproj]: :bulb: When you see these lines appear in the project file, it's safe to manually remove them *if* you are comfortable editing project files by hand. Don't worry about this one; I can remove it when I merge this pull request.

**Comment by sharwell** [comment]
In addition to the items mentioned in the diff:

* :exclamation: Remove the `OpenIssueAttribute` attribute
* :exclamation: You missed one instance of `[TestMethod]` which is causing a compile error

**Comment by sharwell** [comment]
@pdelvo There is no need to update **appveyor.yml**. This pull request is almost ready to merge if you just fix the few items marked with :exclamation: and nothing else. :+1:

**Comment by pdelvo** [comment]
I implemented you the feedback you provided but there are still tests that dont run on my machine when I use the test explorer.

I dont think I missed a TestMethod attribute. I think the build failed because it was doing a merged build which then includes some more TestMethods. Should I rebase the changes on master and fix these or are you fixing them when you merge?

**Comment by sharwell** [comment]
Found the bug in xunit (see the link). Should we do something about it?

**Comment by sharwell** [comment]
I fixed the remaining issue by renaming some conflicting methods.

**Pull request merged**